adobe/copy_on_write.hpp

```diff
@@ -55,93 +36,98 @@ copy when the value is modified and there is more than one reference to the
 value.
 
 \note
-copy_on_write is thread safe when complier with BOOST_HAS_THREADS defined.
+copy_on_write is thread safe.
 
-\note
-To get an instance with a particular allocator one would have to write:
-<pre>
-copy_on_write<T> x(custom_allocator);
-x.write() = value;
-</pre>
 */
-template <typename T,                               // T models Regular
-          typename A = adobe::capture_allocator<T>> // A models Allocator
+template <typename T> // T models Regular
 class copy_on_write {
-public:
-    /// \brief The type of value stored.
-    typedef T value_type;
-    /// \brief The type of allocator.
-    typedef A allocator_type;
-
 #if !defined(ADOBE_NO_DOCUMENTATION)
-private:
-    struct implementation_t;
-    typedef typename allocator_type::template rebind<implementation_t>::other
-    implementation_allocator_type;
 
-public:
+    struct model {
+        std::atomic<std::size_t> _count{1};
+
+        model() = default;
+
+        template <class V>
+        explicit model(V&& x) : _value(std::forward<V>(x)) {}
+
+        T _value;
+    };
+
+    void destroy() {
+        if (_self && (--_self->_count == 0)) delete _self;
+    }
+
+    model* _self;
+
+    template <class U>
+    using disable_copy = std::enable_if_t<!std::is_same<std::decay_t<U>, copy_on_write>::value>*;
+
+    template <typename U>
+    using disable_copy_assign =
+        std::enable_if_t<!std::is_same<std::decay_t<U>, copy_on_write>::value, copy_on_write&>;
 #endif
 
+public:
+    /// \brief The type of value stored.
+    /* [[deprecated]] */ using value_type = T;
+
+    /// \brief The type of value stored.
+    using element_type = T;
+
     /*!
         The first call to the default constructor will construct a default
-        instance of value_type which will be used for subsequent calls to the
+        instance of element_type which will be used for subsequent calls to the
         default constructor. The default instance will be released at exit.
     */
     copy_on_write() {
-        std::call_once(flag_s, init_default);
-        object_m = default_s;
-        object_m->header_m.get().count_m.increment();
-    }
-
-    /*!
-        Initializes the instance with a custom allocator
-    */
-    explicit copy_on_write(const allocator_type& a) : object_m(0) {
-        implementation_allocator_type other_allocator(a);
-
-        object_m = allocate(other_allocator);
+        static model default_s;
+        _self = &default_s;
+        ++_self->_count;
     }
 
     /*!
     \brief Constructs a new copy_on_write object with a value \c x.
 
     \param x A default value to assign to this object
     */
-    copy_on_write(T x) : object_m(allocate_move(0, std::move(x))) {}
+    template <class U>
+    copy_on_write(U&& x, disable_copy<U> = nullptr) : _self(new model(std::forward<U>(x))) {}
```

Comment:
Yes. Allowing emplace construction here would be good. I'll look to fix it (a little tricky with the copy avoidance... because cow<T> is convertible to const T& the constructor needs the guard to keep from sucking up things that should go to the copy_ctor.